File: examples/ffi/projections/python/host.py
import sys
sys.path.insert(0, "target/bindings")
from projection_python import (
    PROJECTION_PYTHON_CLOSED,
    PROJECTION_PYTHON_EXPIRED_VIEW,
    PROJECTION_PYTHON_OK,
    ResourceDocument,
    load,
    project_component,
)
library = load("target/libprojection_python.so")
document = ResourceDocument(library)
assert document.open(bytes((10, 20, 30))) == PROJECTION_PYTHON_OK
status, stale = document.bytes()
assert status == PROJECTION_PYTHON_OK
assert document.at(stale, 1) == (20, PROJECTION_PYTHON_OK)
assert document.replace(bytes((40, 50, 60))) == PROJECTION_PYTHON_OK
assert document.at(stale, 1) == (0, PROJECTION_PYTHON_EXPIRED_VIEW)
status, fresh = document.bytes()
assert status == PROJECTION_PYTHON_OK
assert document.at(fresh, 1) == (50, PROJECTION_PYTHON_OK)
assert document.close() == PROJECTION_PYTHON_OK
assert document.close() == PROJECTION_PYTHON_CLOSED
assert library.label("hello") == "hello"
assert project_component(library, {"count": 7, "label": "ok"}) == {"count": 7, "label": "ok"}
assert library.add(41) == 42

File: examples/ffi/projections/javascript/host.mjs
import {
    ResourceDocument,
    RESOURCE_OK,
    RESOURCE_EXPIRED_VIEW,
    RESOURCE_CLOSED,
    add,
    label,
    project_component,
    load,
} from "./target/bindings/projection_javascript.mjs";

const library = load(process.env.JET_NODE_ADDON ?? "./target/libprojection_javascript.node");
const document = new ResourceDocument(library);
if (document.open(Uint8Array.from([10, 20, 30])) !== RESOURCE_OK) process.exitCode = 1;
const staleResult = document.bytes();
if (staleResult.status !== RESOURCE_OK) process.exitCode = 1;
const initialAt = document.at(staleResult.view, 1);
if (initialAt.status !== RESOURCE_OK || initialAt.value !== 20n) process.exitCode = 1;
if (document.replace(Uint8Array.from([40, 50, 60])) !== RESOURCE_OK) process.exitCode = 1;
if (document.at(staleResult.view, 1).status !== RESOURCE_EXPIRED_VIEW) process.exitCode = 1;
const freshResult = document.bytes();
if (freshResult.status !== RESOURCE_OK || document.at(freshResult.view, 1).value !== 50n) process.exitCode = 1;
if (document.close() !== RESOURCE_OK || document.close() !== RESOURCE_CLOSED) process.exitCode = 1;
if (add(library, 41n) !== 42n) process.exitCode = 1;
if (label(library, "hello").toString("utf8") !== "hello") process.exitCode = 1;
const projected = project_component(library, { count: 7, label: "ok" });
if (projected.count !== 7 || projected.label !== "ok") process.exitCode = 1;
